### PR TITLE
Migrated ZimHostFragment -> ZimHostRoute

### LINKED ISSUE #4615: Remove `ZimHostFragment` from project, and use pure composable instead.
Parent Issue https://github.com/kiwix/kiwix-android/issues/4572

### PR DESCRIPTION
**Pull request by atharvyadav22**
Fixes: #4615 

Additionally added/modified complete End-To-End Testing Setup
- Added Unit test for ZimHostViewModel
- Added UI Test for ZimHostScreen
- Refactor testing setup according to new composable pattern.

### CODE REVIEW

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
Sorry for the delay and not able to contribute in other Prs i was prioritizing this migration it was difficult because of many classes which required understanding.
This is an initial commit ill later add test cases too for now please review the logic.

I have added comments to help other read the code, also from previous migration i have implemented your points so it saves time.

Migration covered these parts-
- Hotspot now use composable lifecycle
- Merged Di and common fragment files and updated in ViewModel itself
- Handled Playstore and custom apps logic 
- Permission block kept the same
- Instead of suppressing i made small code blocks.
- for the checkbox i modified the logic cause it wasn't updating its state with toggleSelection.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz Done with this migration waiting for your reviews.

File: app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt [hunk 1]
```diff
@@ -68,6 +68,7 @@ const val BOOK_ITEM_TESTING_TAG = "bookItemTestingTag"
 fun BookItem(
   index: Int,
   bookOnDisk: BookOnDisk,
+  isSelected: Boolean = bookOnDisk.isSelected,
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt]: Why are we adding this parameter? It can be easily used from the bookOnDisk object. It is unnecessary to add a new parameter everywhere when we can simply use it from the object that is already passed in the used methods.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostScreen.kt [hunk 2]
```diff
@@ -82,6 +82,7 @@ fun ZimHostScreen(
   booksList: List<BooksOnDiskListItem>,
   startServerButtonItem: Triple<String, Color, () -> Unit>,
   selectionMode: SelectionMode,
+  selectedBookIds: Set<String>,
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostScreen.kt]: Instead of this, we can directly update this in booksList since it type of `BooksOnDiskListItem` that has the `isSelected` property.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+  CollectZimHostEvents(
+    lifecycleOwner,
+    viewModel,
+    activity,
+    alertDialogShower
+  )
+
+  val serverIpDisplayText =
+    if (uiState.isServerRunning && uiState.ipAddress.isNotBlank()) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: The logic part should be inside the viewModel. Compose should only be for rendering it.

Like: This code should be inside the viewModel in a method, maybe `serverIpText`, and should be called in UI directly `viewModel.serverIpText()`.

```kotlin
if (uiState.isServerRunning && uiState.ipAddress.isNotBlank()) {
      stringResource(id = R.string.server_started_message, uiState.ipAddress)
    } else {
      stringResource(id = R.string.server_textview_default_message)
    }
```

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+      stringResource(id = R.string.server_started_message, uiState.ipAddress)
+    } else {
+      stringResource(id = R.string.server_textview_default_message)
+    }
+
+  val buttonText =
+    if (uiState.isServerRunning) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: Same for this.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+      stringResource(id = R.string.stop_server_label)
+    } else {
+      stringResource(id = R.string.start_server_label)
+    }
+
+  val buttonColor =
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: Same for this.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    if (uiState.isServerRunning) StopServerRed else StartServerGreen
+
+  ZimHostScreen(
+    serverIpText = serverIpDisplayText,
+    shareIconItem = uiState.shareVisible to {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: We can improve this lambda here. Currently, it is recreated on every recomposition.  Not a huge issue, but it is unneccessary creating this multiple times without a good reason.

It should be something like:

```kotlin
val shareIconClick = remember(uiState.ipAddress) {
  {
    activity.startActivity(
      Intent(Intent.ACTION_SEND).apply {
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, uiState.ipAddress)
      }
    )
  }
}
```

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+// ===== Permission Handling =====
+@Suppress("ReturnCount")
+private fun handlePermissionsAndStart(
+  activity: KiwixMainActivity,
+  uiState: ZimHostViewModel.UiState,
+  alertDialogShower: AlertDialogShower,
+  viewModel: ZimHostViewModel
+) {
+  // Handles Android 13+ Notification Permission
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: We have the `KiwixPermissionChecker`; we should use that. You should take reference from the already used places. It will help us to maintain the permission checking at one place.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    val granted = ContextCompat.checkSelfPermission(
+      activity, Manifest.permission.POST_NOTIFICATIONS
+    ) == PackageManager.PERMISSION_GRANTED
+
+    if (!granted) {
+      ActivityCompat.requestPermissions(
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: If the permission is not granted, we should ask for it by `rememberLauncherForActivityResult`, see the `SettingsScreen` implementation, how we are checking for permission, and if not granted, then how we are asking for the permission in UI via events.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        activity,
+        arrayOf(Manifest.permission.POST_NOTIFICATIONS),
+        REQUEST_POST_NOTIFICATION_PERMISSION
+      )
+      return
+    }
+  }
+
+  // Handles Play Store
+  if (uiState.isPlayStoreBuild) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: this uiState is confusing. It is telling us it is `playStore` build. But it is playStore build with Android 11 and above.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    viewModel.onButtonClicked()
+    return
+  }
+
+  // Handles Android 11+ Manage External Storage (Non-Custom Apps)
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R &&
+    !activity.isCustomApp() &&
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: Why are we not asking this permission in custom apps?

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    !Environment.isExternalStorageManager()
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: Do not directly check it by this method. See `Activity.isManageExternalStoragePermissionGranted` method. Include that method in KiwixPermissionChecker and use that.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+        activity.startActivity(
+          Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
+        )
+      }
+    )
+    return
+  }
+
+  // Handles Android 10 below Storage permission
+  if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: Use this same way as we are using in settingsScreen.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+    val granted = ContextCompat.checkSelfPermission(
+      activity, Manifest.permission.READ_EXTERNAL_STORAGE
+    ) == PackageManager.PERMISSION_GRANTED
+
+    if (!granted) {
+      ActivityCompat.requestPermissions(
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostRoute.kt]: The other issue, you are only asking for permission, not handling the permission denied/granted scenarios. If a user is granted the permission, then they have to click again on the button to start the server, which is a wrong UX. You should handle them like we were handling them in the fragment code.

I suggest you to go through the `SettingsScreen` codebase you will have a better idea about this.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+      if (!isCustomApp && hosted.isEmpty() && initialSelection.isNotEmpty()) {
+        kiwixDataStore.setHostedBookIds(initialSelection)
+      }
+
+      if (ServerUtils.isServerStarted) onServerStarted(ServerUtils.serverAddress) else layoutStopped()
+    }
+  }
+
+  fun onButtonClicked() {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostViewModel.kt]: The method should improve. It should be `startServerButtonClick`.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@atharvyadav22 Please see the review changes. If anything is unclear, please ask me before implementing it. Since this PR has many logic changes. It takes significant time in testing/reviewing PRs like this one.

**Comment by atharvyadav22** [comment]
> @atharvyadav22 Please see the review changes. If anything is unclear, please ask me before implementing it. Since this PR has many logic changes. It takes significant time in testing/reviewing PRs like this one.

Yes thanks for your valuable time and pointing out i always try to not many many possible errors and solve then push commits but migration is lengthy anyways i'll take care of in in next commit i'll solve everything you mentioned so it won't take your time.
Thanks for detailed review ❤️

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Mark your PR as a draft when the effort is ongoing.

File: app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt (hunk 1 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt]: @MohitMaliFtechiz the issue was of State updating of the checkbox like when before starting that Hotspot State our checkbox don't updates so that fix was placed without disturing other files.
I have mentioned that in Pr description earlier.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz i'll prioritize this one issue now.

File: app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt (hunk 1 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt]: @atharvyadav22 The state can be updated without this. I have made the changes in the current code. You can see them. Some more changes are left, which I am working on. After that, I will ask you to write the test cases :).

Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt]: should i do that review changes and test cases so it help you look other migration work? like if there is nothing much to do other than review changes.

Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/ui/BookItem.kt]: No, there are not only the review changes. I have found some major issues here(Not due to this PR code, but in general). So i am fixing that one also. I am pushing the changes. After that, the permission improvement is left do it. Then you can start writing the test cases.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz i have checked the code and functionality it works fine for all the cases.
Coming up to test cases there are concerning issues lie.
1. flakyTest{} wrapper
2. testFlow should be replaced with test{} (Kotlin has this)
3. Turbine should be closed 
4. 
```
@Suppress("InjectDispatcher")
  private var ioDispatcher: CoroutineDispatcher = Dispatchers.IO
```
this should be passed via Constructor, also for my PR of stabilization IO i didn't touched this as this is little complex and as migration was going.

similarly for this
```
  fun sendUiEvent(uiEvent: OnlineLibraryUiEvent) =
    viewModelScope.launch {
      _onlineLibraryEvent.emit(uiEvent)
    }
```
we cant control this in tests

so to properly make an robust test we need to consider this after migration properly i suppose, cause we have seen many issues with these in past also.

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Now, this PR code is set. You can start writing the test cases(unit test cases) for ZimHostViewModel and other things introduced in this PR. If any UI test case fails, please have a look at that as well :)

If you need anything from me, just drop a comment here.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz okay ill do it here i see new pushes 
Thanks for your help 👍

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 For `sendUiEvent` method. I have added a comment on that method. This will be updated in the `OnlineLibraryScreen` migration. 

Here we need to add/update/improve the test cases for this PR. Not for other screens. Since `ZimManageViewModel` is very large class and handles 2 screen datas. We need to first split them into seperate viewModels, and then improve those test cases.

**Comment by atharvyadav22** [comment]
> @atharvyadav22 For `sendUiEvent` method. I have added a comment on that method. This will be updated in the `OnlineLibraryScreen` migration.
> 
> Here we need to add/update/improve the test cases for this PR. Not for other screens. Since `ZimManageViewModel` is very large class and handles 2 screen datas. We need to first split them into seperate viewModels, and then improve those test cases.

new tests file for newly created methods right only or any specific things?

As we are in urgent for migration for that `OnlineLibraryScreen` i see no progress as that PR has many issues including many CI failure, if you allow i can take that : ) since my migration related tasks are almost completed. 
between 1-1.5 week i can complete that migration.
Thanks.

**Comment by MohitMaliFtechiz** [comment]
> new tests file for newly created methods right only or any specific things?

The newly created files with edge cases.

> As we are in urgent for migration for that OnlineLibraryScreen i see no progress as that PR has many issues including many CI failure, if you allow i can take that : ) since my migration related tasks are almost completed.
between 1-1.5 week i can complete that migration.
Thanks.

Thank you for showing interest. But OnlineScreen has many logics and scenarios. I will take that. You can take some other issues  :).

**Comment by atharvyadav22** [comment]
> > new tests file for newly created methods right only or any specific things?
> 
> The newly created files with edge cases.
> 
> > As we are in urgent for migration for that OnlineLibraryScreen i see no progress as that PR has many issues including many CI failure, if you allow i can take that : ) since my migration related tasks are almost completed.
> > between 1-1.5 week i can complete that migration.
> > Thanks.
> 
> Thank you for showing interest. But OnlineScreen has many logics and scenarios. I will take that. You can take some other issues :).

Okay if there is anything which might help kiwix and save your time i can help with that if something you feel.
Thanks.

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Any update here?

**Comment by atharvyadav22** [comment]
> @atharvyadav22 Any update here?

Yes actually i am doing coverage also side by side for this. I am currently doing that Media Save PR as the recent test files fails for new code so its taking me time to rewrite them along side i am doing that `UnsupportedMimeType` as you said so.

I'll do this PR by this week itself, actually i am using best practice and trying to make test robust so it saves your reviewing time to so its taking me time.
Thanks.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
This is only for Unit Test cases but inorder to make an robust migration i'll also make Instrument Test like i have started working on it should i go on ?

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Yes, we should add both instrumentation and unit test cases. However, the UI test cases are already exist. Those are the scenarios based. But you can add the compose UI test cases as well. So that we can verify the UI as well.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
actually the Current UI Test is not that good reason is we need to mimic actual behavior of the Screen like server starting, etc.

Now interesting thing is this can't work on CI as emulators dont have these _WIFI_ but if we test on real device rather it should work i came up with idea like we'll check about _WIFI_ to know whether we are testing in emulator or real devices also will be placing a fix to run for all the devices API 27+ (current has upto 32 i suppose)   

I'll need time for such things as for an robust migration.

Thanks.

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Previously, we had the fragments, and we could not test the functionality directly within a fragment for real scenarios. But now we have migrated to compose with view models.

So now we can test layer by layer to ensure all the functionality works correctly.

* First layer should be the viewModel testing, where all the logic is written. By doing this, we can ensure how this functionality will behave on an actual device.
* The second layer is the UI layer. A pure compose UI testing, so that we can ensure which data come from the viewModel is rendered correctly and shows the correct UI. Also, when clicks are performed, they should call the correct viewModel methods(ViewModel testing is already done).
* The third one is a pure real-scenario test that is already written. Due to the unavailability of WIFI on emualtor it is currently running on the API level 30, which is okay because we have tested both UI and logic above the testing layer.

By doing this, we don't need to add additional testing and conditions based on emualtor or real device.

**Comment by atharvyadav22** [comment]
> @atharvyadav22 Previously, we had the fragments, and we could not test the functionality directly within a fragment for real scenarios. But now we have migrated to compose with view models.
> 
> So now we can test layer by layer to ensure all the functionality works correctly.
> 
> * First layer should be the viewModel testing, where all the logic is written. By doing this, we can ensure how this functionality will behave on an actual device.
> * The second layer is the UI layer. A pure compose UI testing, so that we can ensure which data come from the viewModel is rendered correctly and shows the correct UI. Also, when clicks are performed, they should call the correct viewModel methods(ViewModel testing is already done).
> * The third one is a pure real-scenario test that is already written. Due to the unavailability of WIFI on emualtor it is currently running on the API level 30, which is okay because we have tested both UI and logic above the testing layer.
> 
> By doing this, we don't need to add additional testing and conditions based on emualtor or real device.

Okay actually what i was working was if someone tests on real devices they will see that navigation, clicks eg- Server is started, books selection toggle.
Keeping CI safe.

I also agree with your point so should i continue with that work or only fix that CI errors?

**Comment by MohitMaliFtechiz** [comment]
We should add the testing as I described in layers. That will already ensure that everthing works correctly. Devs can easily run the test on the required API level. However, if there are some changes made in this screen, then devs first test manually. So we should not add conditions here.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 

Sorry forgot to `amend `the commit

Fixed Ci for Instrumentation Test and Renamed as per Layer 3 (Integration Test)
Added Test for layer 2 (UI Test)
Address review changes 

Also Apologies for bad commit the reason is there was merge conflict so had to reset head .

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Thanks for the update.

There are still some conflicts in this PR. Also, in this commit https://github.com/kiwix/kiwix-android/pull/4736/changes/1c778bfbb022ace7f8166487680475fcf8af4e04 there are still some unnecessary messages in the commit; remove that.


image

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
Could you verify everything is fine now any improvements?
also there were many merge conflicts i have resolved them still please verify if i did correctly the rebase part for files like `KiwixFilePermissionChecker` and others. 
I have rewritten the history as well and edited the commit message.
Thanks.

**Comment by MohitMaliFtechiz** [comment]
For asking permission, use the `accompanist` library. See #4819.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt [hunk 15]
```diff
@@ -139,9 +144,10 @@ class ZimHostFragmentTest {
   }
 
   @Test
-  fun testZimHostFragment() {
+  fun testZimHostScreen() {
     if (Build.VERSION.SDK_INT > Build.VERSION_CODES.N_MR1 &&
-      Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
+      Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU &&
+      isWifiEnabled()
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt]: Good improvement 👏 . I guess now we should remove other conditions and only place `isWifiEnabled()` condition.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt [hunk 16]
```diff
@@ -242,11 +254,11 @@ class ZimHostFragmentTest {
 
   private fun loadZimFileInApplication(zimFileName: String) {
     val loadFileStream =
-      ZimHostFragmentTest::class.java.classLoader.getResourceAsStream(zimFileName)
+      ZimHostScreenInstrumentTest::class.java.classLoader!!.getResourceAsStream(zimFileName)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt]: Instead of the double bang operator, we should use `Safe Call Operator` (?) here.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+import org.kiwix.kiwixmobile.core.ui.theme.StartServerGreen
+import org.kiwix.kiwixmobile.core.ui.theme.StopServerRed
+import org.kiwix.kiwixmobile.core.utils.ZERO
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.BooksOnDiskListItem.BookOnDisk
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.SelectionMode
+import org.kiwix.kiwixmobile.ui.BOOK_ITEM_CHECKBOX_TESTING_TAG
+import java.io.File
+
+@RunWith(AndroidJUnit4::class)
+class ZimHostScreenTest {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: Here we should run this with `Robolectric`. Also, we should test the `onMultiSelect` click event as well.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostScreen.kt
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostScreen.kt]: Here we should use a compose-based dimension instead of an XML-based. See https://github.com/kiwix/kiwix-android/pull/4823/changes#diff-58034fc6348dc88af4f04bb94175491a65e43675f3e8b644d013536aa1e3dd1fR176 how we have done this.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@atharvyadav22 Please check why `testZimHostScreen` is failing, and fix that.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt (hunk 17 above)
Comment (atharvyadav22) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: Yes, i thought the same but issue is this uses real hardware like Hotspot and Wifi to work so real device is more appropriate. what say?

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt (hunk 16 above)
Comment (atharvyadav22) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenInstrumentTest.kt]: okay i'll use an error message too for that making it clear

File: app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt (hunk 17 above)
Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: @atharvyadav22 It is a screen test, and we are verifying only whether the UI part is working correctly or not. Rendering the state, click events, etc. Here, we are not testing running a hotspot. For that, we already have the other test.

Comment (atharvyadav22) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: @MohitMaliFtechiz 
I tried that but

```
Could not find 'c++_shared.dll'. Looked for: [armeabi-v7a], but only found: [java.util.regex.PatternSyntaxException: Unmatched closing ')' near index 10
lib\([^\\]*)\c++_shared.dll
          ^].
com.getkeepsafe.relinker.MissingLibraryException: Could not find 'c++_shared.dll'. Looked for: [armeabi-v7a], but only found: [java.util.regex.PatternSyntaxException: Unmatched closing ')' near index 10
lib\([^\\]*)\c++_shared.dll
```

then i came accross this have a look https://github.com/robolectric/robolectric/issues/6123#issuecomment-854719201 which says Native JNI isnt supported and our LibKiwix is used in the test 

while this says https://stackoverflow.com/questions/68164339/error-with-tests-in-android-studio-when-using-relinker-to-load-libraries remove library

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: @atharvyadav22 Libkiwix will not work in Robolectric(we know that). But where is the Libkiwix used in this test? It is a pure UI test that depends on the provided clickListener, qrItem, etc.

Comment (atharvyadav22) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: @MohitMaliFtechiz 
I tried debugging but still same issue from my side don't know like for Instrumention Test it worked when i change context and scenerio for Roboelectric it gives me this error maybe windows platform issue or something i am missing out can you have a look on it please.
Thanks.

Comment (MohitMaliFtechiz) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: Okay sure.

Comment (atharvyadav22) [reply app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostScreenTest.kt]: Other than this issue everything is solved. 
Thanks

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
I have solve the error it was very silly i forgot to add `application = TestApplication::class`.
have completed all the work, waiting if something is left.
Thanks

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 Thank you. I will have a look at this after https://github.com/kiwix/kiwix-android/pull/4689.

**Comment by atharvyadav22** [comment]
sorry accidential click

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
Anything more i should add here?

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22 I have not looked at this yet. I am busy making the architecture of https://github.com/kiwix/kiwix-android/pull/4689. That is a very important part of our application. I will have a look after that.

**Review by MohitMaliFtechiz — DISMISSED** [review]
@atharvyadav22 Thank you. @kelson42 Now, LGTM. Once CI is green, we can merge.

**Comment by atharvyadav22** [comment]
Looks like we should keep the test disabled only and fix them seperately

**Review by MohitMaliFtechiz — DISMISSED** [review]

**Comment by kelson42** [comment]
@MohitMaliDeveloper Sorry, but can you please rebase?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 done.
